### PR TITLE
Loop invariant motion

### PR DESCRIPTION
**Pull request by clyben**
This PR add loop invariant motion.
In a Theta node, an operand is movable out of Theta when 
1. It's an invariant operand, and is on boundary of invariant and non-invariant. (e.g. op1 is invariant but (+ op1 op2) is not invariant, then op1 is on boundary)
2. The operand is not basic input like (Arg n), or otherwise it's not needed to move.
Currently it's facing performance issues which are likely caused by substitution. It might be solved with must-subst-beneath.

### CODE REVIEW

**Comment by clyben** [comment]
currently it's wouldn't generate the wanted code due to the OperandGroup extraction problem.

File: src/rvsdg/egglog_optimizer/loop-optimizations.egg [hunk 1]
```diff
@@ -1,12 +1,15 @@
-
+(function unroll_count (i64) i64 :unextractable :merge (max old new))
```
Comment (oflatt) [inline src/rvsdg/egglog_optimizer/loop-optimizations.egg:1]: Do you still need this?

**Review by oflatt — CHANGES_REQUESTED** [review]
Had a quick question

File: src/rvsdg/egglog_optimizer/loop-optimizations.egg (hunk 1 above)
Comment (clyben) [reply src/rvsdg/egglog_optimizer/loop-optimizations.egg:1]: yes, setting the unroll max to 2 will make the test not finishing..

Comment (oflatt) [reply src/rvsdg/egglog_optimizer/loop-optimizations.egg:1]: Okay, fix after today!

**Review by oflatt — APPROVED** [review]
Great work!

**Pull request merged**